### PR TITLE
KAFKA-13518: Update gson dependency

### PR DESCRIPTION
**Pull request by dongjinleekr**
Here is the fix. Since [spotbugs 4.5.1 was released just 12 hours ago](https://github.com/spotbugs/spotbugs/releases/tag/4.5.1), it would take a little bit to be synched with maven central.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by showuon — COMMENTED** [review]
@dongjinleekr , thanks for the PR. But there's spotBugs 4.5.2. Could you upgrade to the latest one? Also, the PR title/description can't tell what's the relationship between spotBugs and gson, could you add explanation there? Thanks.

**Review by showuon — COMMENTED** [review]
Also, there are build errors. Please fix them. Thanks.

**Comment by dongjinleekr** [comment]
Hi @showuon,

Here it is. I just updated the issue title more clearly and updated the spotbugs dependency into 4.5.2.

**Review by showuon — APPROVED** [review]
LGTM!

**Review by showuon — COMMENTED** [review]
Please also help fix the build failure. Thanks.

**Comment by dongjinleekr** [comment]
@showuon Sorry for bothering. Here is the update. There were some updates on spotbugs between 4.2.2 and 4.5.2:

- In 4.3.0, spotbugs improved their detection logic to detect :

**Comment by dongjinleekr** [comment]
@showuon Sorry for bothering you. Here is the update. There were some [updates on spotbugs between 4.2.2 and 4.5.2](https://github.com/spotbugs/spotbugs/blob/master/CHANGELOG.md) and some previously-unfound problems are now detected:

- In 4.3.0, spotbugs improved their detection logic for `MS_EXPOSE_REP`: 

  > `MS_EXPOSE_REP` and `EI_EXPOSE_REP` are now reported for code returning a reference to a mutable object indirectly (e.g. via a local variable)

- In 4.4.2, spotbugs fixed some false positives for `DMI_RANDOM_USED_ONLY_ONCE` and started to detect some unfound problems:

  > `DMI_RANDOM_USED_ONLY_ONCE` false positive

After the update, it works like a charm:

**Comment by dongjinleekr** [comment]
retest this please.

**Review by showuon — COMMENTED** [review]
@dongjinleekr , thanks for the update, but it looks like there are other failed spotBugs.

**Comment by dongjinleekr** [comment]
@showuon My bad. I found several other false-positives from other modules with spotbugs; They are now fixed. (Please see the comments.) :bow:

**Comment by ijuma** [comment]
Thanks for the PR. Seems like the new version has more false positives. Do you know if they intend to fix those?

**Comment by dongjinleekr** [comment]
@ijuma

> Do you know if they intend to fix those?

Oh yes, as you can see in the updated PR, I updated spotbugs to 4.5.3 following the gradle plugin 5.0.5, and rebased onto the latest trunk. It seems like there are a bunch of false positives in the recent version of spotbugs (below) but, I verified that none of them are affecting.



+1. They also have not fixed the issues I commented on in `spotbugs-exclude.xml` yet. I will follow up and apply them as soon as they fix them.

**Comment by ijuma** [comment]
Should we wait until they fix these issues in spotBugs? It doesn't look like the cost/benefit in upgrading here isn't favorable.

**Comment by dongjinleekr** [comment]
@ijuma If you don't mind [CVE WS-2021-0419](https://github.com/protocolbuffers/protobuf/issues/9457) introduced by gson 2.8.6. This PR is to fix it.

**Review by andrewchoi5 — APPROVED** [review]

**Comment by dongjinleekr** [comment]
Rebased onto the latest trunk. cc/ @ijuma

**Comment by Boojapho** [comment]
Gson library has another recent vulnerability: https://nvd.nist.gov/vuln/detail/CVE-2022-25647.  Gson library 2.8.9 fixes this, which is included in Spotbugs 4.5.0 and higher.

**Comment by dongjinleekr** [comment]
@Boojapho Thanks for reporting. Here is the fix - rebased onto the latest trunk and upgraded spotbugs into 4.7.0, which also fixes the gson vulnerability.

File: gradle/spotbugs-exclude.xml [hunk 1]
```diff
@@ -105,6 +127,19 @@ For a detailed description of spotbugs bug categories, see https://spotbugs.read
         <Bug pattern="NP_NONNULL_PARAM_VIOLATION"/>
     </Match>
 
+    <!-- false positive, see:
+       https://github.com/spotbugs/spotbugs/issues/600
+       https://github.com/spotbugs/spotbugs/issues/1791
```
Comment (ijuma) [inline gradle/spotbugs-exclude.xml:129]: Looks like https://github.com/spotbugs/spotbugs/commit/720af6cdd592d2e49d3a472465eed9ad361ebd27 would help, but it's not available in a released version yet.

File: gradle/spotbugs-exclude.xml [hunk 2]
```diff
@@ -43,6 +54,17 @@ For a detailed description of spotbugs bug categories, see https://spotbugs.read
             EQ_COMPARETO_USE_OBJECT_EQUALS: This class defines a compareTo method but no equals() method -->
         <Bug pattern="EQ_COMPARETO_USE_OBJECT_EQUALS"/>
     </Match>
+    <Match>
+        <!-- Disable bug types introduced by the ThrowingExceptions detector with 4.7.0.
```
Comment (ijuma) [inline gradle/spotbugs-exclude.xml]: Disabled by default here https://github.com/spotbugs/spotbugs/commit/7ba0e74fae9390662be88a2ee9a51090af28f8c0 (not released yet).

File: gradle/spotbugs-exclude.xml [hunk 3]
```diff
@@ -28,8 +28,19 @@ For a detailed description of spotbugs bug categories, see https://spotbugs.read
         <!-- Disable warnings about mutable objects and the use of public fields.
             EI_EXPOSE_REP: May expose internal representation by returning reference to mutable object
             EI_EXPOSE_REP2: May expose internal representation by incorporating reference to mutable object
+            MS_EXPOSE_REP: Public static method may expose internal representation by returning array
```
Comment (ijuma) [inline gradle/spotbugs-exclude.xml]: Why do we have to disable this one?

File: gradle/spotbugs-exclude.xml [hunk 4]
```diff
@@ -28,8 +28,19 @@ For a detailed description of spotbugs bug categories, see https://spotbugs.read
...
+            <Bug pattern="EI_EXPOSE_REP"/>
+            <Bug pattern="EI_EXPOSE_REP2"/>
+            <Bug pattern="MS_EXPOSE_REP"/>
+            <Bug pattern="MS_PKGPROTECT"/>
+        </Or>
+    </Match>
+    <Match>
+        <!-- Disable warnings about Random object.
+            DMI_RANDOM_USED_ONLY_ONCE: Random object created and used only once -->
+        <Bug pattern="DMI_RANDOM_USED_ONLY_ONCE"/>
```
Comment (ijuma) [inline gradle/spotbugs-exclude.xml]: Are we disabling this one due to false positives?

File: gradle/spotbugs-exclude.xml (hunk 3 above)
Comment (dongjinleekr) [reply gradle/spotbugs-exclude.xml]: 1. Public static `org.apache.kafka.server.metrics.KafkaYammerMetrics.defaultRegistry()` may expose internal representation by returning `KafkaYammerMetrics.metricsRegistry`.
2. Public static `org.apache.kafka.common.security.auth.SecurityProtocol.names()` may expose internal representation by returning `SecurityProtocol.NAMES`

File: gradle/spotbugs-exclude.xml (hunk 4 above)
Comment (dongjinleekr) [reply gradle/spotbugs-exclude.xml]: Yes.

1. Random object created and used only once in new `kafka.tools.ConsoleConsumer$ConsumerConfig(String[])`
2. Random object created and used only once in new `kafka.tools.ConsumerPerformance$ConsumerPerfConfig(String[])`

**Comment by dongjinleekr** [comment]
@ijuma Here is the update:

1. Rebased onto the latest trunk.
2. Gather the false positives together and add some TODO comments not to leave the workarounds later.
3. Reduce the scope of `DMI_RANDOM_USED_ONLY_ONCE` and `MS_EXPOSE_REP` into scala-only and some specific classes only, respectively.

**Comment by ijuma** [comment]
Can you please update to spotbugs 4.7.1? It seems like it fixes the false positives.

**Comment by omkreddy** [comment]
@dongjinleekr I have not seen this PR. I have raised draft PR https://github.com/apache/kafka/pull/12768 to upgrade spotbugs. Can we update to spotbugs 4.7.3?

**Comment by omkreddy** [comment]
Closing this PR in favour of https://github.com/apache/kafka/pull/12768

**Pull request closed without merging**